File: lib/src/common/cookie.dart
enum CookieSameSiteStatus {
  NO_RESTRICTION("no_restriction"),
  LAX("lax"),
  STRICT("strict"),
  UNSPECIFIED("");

  const CookieSameSiteStatus(this.value);

  final String value;
}

class CookiePartitionKey {
  String topLevelSite;

  CookiePartitionKey(this.topLevelSite);

  CookiePartitionKey.fromMap(Map<Object?, Object?> map)
      : topLevelSite = map["topLevelSite"] as String;

  Map<String, dynamic> toMap() => {
    "topLevelSite": topLevelSite
  };
}

class Cookie {
  String domain;
  int? expirationDate;
  String firstPartyDomain;
  bool hostOnly;
  bool httpOnly;
  String name;
  CookiePartitionKey? partitionKey;
  String path;
  bool secure;
  bool session;
  CookieSameSiteStatus sameSite;
  String storeId;
  String value;


  Cookie(
      this.domain,
      this.expirationDate,
      this.firstPartyDomain,
      this.hostOnly, this.httpOnly,
      this.name,
      this.partitionKey,
      this.path,
      this.secure,
      this.session,
      this.sameSite,
      this.storeId,
      this.value);

  Cookie.fromMap(Map<Object?, Object?> map)
      : domain = map["domain"] as String,
        expirationDate = map["expirationDate"] as int?,
        firstPartyDomain = map["firstPartyDomain"] as String,
        hostOnly = map["hostOnly"] as bool,
        httpOnly = map["httpOnly"] as bool,
        name = map["name"] as String,
        partitionKey = ((map["partitionKey"] as Map<Object?, Object?>?) != null ?
        CookiePartitionKey.fromMap((map["partitionKey"] as Map<Object?, Object?>?)!)
            : null),
        path = map["path"] as String,
        secure = map["secure"] as bool,
        session = map["session"] as bool,
        sameSite = CookieSameSiteStatus.values
            .firstWhere((element) => element.value == map["sameSite"] as String),
        storeId = map["storeId"] as String,
        value = map["value"] as String;

  Map<String, dynamic> toMap() => {
    "domain": domain,
    "expirationDate": expirationDate,
    "firstPartyDomain": firstPartyDomain,
    "hostOnly": hostOnly,
    "httpOnly": httpOnly,
    "name": name,
    "partitionKey": partitionKey?.toMap(),
    "path": path,
    "secure": secure,
    "session": session,
    "sameSite": sameSite.value,
    "storeId": storeId,
    "value": value
  };
}

File: android/src/main/kotlin/info/xert/gecko_view_flutter/common/Cookie.kt
package info.xert.gecko_view_flutter.common

import org.json.JSONObject

enum class CookieSameSiteStatus(val value: String) {
    NO_RESTRICTION("no_restriction"),
    LAX("lax"),
    STRICT("strict"),
    UNSPECIFIED("")
}

data class CookiePartitionKey(
        val topLevelSite: String
) : OutputStructure, JSONOutputStructure {
    override fun toMap(): Map<String, Any?> {
        val map = mutableMapOf<String, Any?>()

        map["topLevelSite"] = topLevelSite

        return map
    }

    override fun toJSON(): JSONObject {
        val json = JSONObject()

        json.put("topLevelSite", topLevelSite)

        return json
    }

    companion object: JSONInputStructure<CookiePartitionKey>, InputStructure<CookiePartitionKey> {
        override fun fromJSON(inputJSON: JSONObject): CookiePartitionKey {
            return CookiePartitionKey(
                    topLevelSite = inputJSON.getString("topLevelSite")
            )
        }
        override fun fromMap(inputMap: Map<*, *>): CookiePartitionKey {
            return CookiePartitionKey(
                    topLevelSite = inputMap["topLevelSite"] as String
            )
        }
    }
}

data class Cookie(
        val domain: String,
        val expirationDate: Int?,
        val firstPartyDomain: String,
        val hostOnly: Boolean,
        val httpOnly: Boolean,
        val name: String,
        val partitionKey: CookiePartitionKey?,
        val path: String,
        val secure: Boolean,
        val session: Boolean,
        val sameSite: CookieSameSiteStatus,
        val storeId: String,
        val value: String
) : OutputStructure {
    override fun toMap(): Map<String, Any?> {
        val map = mutableMapOf<String, Any?>()

        map["domain"] = domain
        map["expirationDate"] = expirationDate
        map["firstPartyDomain"] = firstPartyDomain
        map["hostOnly"] = hostOnly
        map["httpOnly"] = httpOnly
        map["name"] = name
        map["partitionKey"] = partitionKey?.toMap()
        map["path"] = path
        map["secure"] = secure
        map["session"] = session
        map["sameSite"] = sameSite.value
        map["storeId"] = storeId
        map["value"] = value

        return map
    }

    companion object: JSONInputStructure<Cookie> {
        override fun fromJSON(inputJSON: JSONObject): Cookie {
            val partitionKeyRaw = toValueOrNull(inputJSON, "partitionKey") as JSONObject?
            val partitionKey = if (partitionKeyRaw == null || partitionKeyRaw.length() == 0) null
                               else CookiePartitionKey.fromJSON(partitionKeyRaw)

            return Cookie (
                    domain = inputJSON.getString("domain"),
                    expirationDate = toValueOrNull(inputJSON, "expirationDate") as Int?,
                    firstPartyDomain = inputJSON.getString("firstPartyDomain"),
                    hostOnly = inputJSON.getBoolean("hostOnly"),
                    httpOnly = inputJSON.getBoolean("httpOnly"),
                    name = inputJSON.getString("name"),
                    partitionKey = partitionKey,
                    path = inputJSON.getString("path"),
                    secure = inputJSON.getBoolean("secure"),
                    session = inputJSON.getBoolean("session"),
                    sameSite = CookieSameSiteStatus.entries.first { status -> status.value == inputJSON.getString("sameSite") },
                    storeId = inputJSON.getString("storeId"),
                    value = inputJSON.getString("value")
            )
        }
    }
}
